### PR TITLE
Conv2D forward without loops much faster, backward slightly faster

### PR DESCRIPTION
**Pull request by marcelbischoff**

### CODE REVIEW

**Comment by geohot** [comment]
Woah! Can you add some comments explaining how it works?

File: tinygrad/ops.py [hunk 1]
```diff
@@ -163,16 +163,15 @@ def forward(ctx, x, w, stride=1, groups=1):
...
-          ret[:, g*rcout:(g*rcout+rcout), Y, X] += tx.dot(tw)
-    return ret
+    gx = x.reshape(bs,ctx.groups,cin,x.shape[2],x.shape[3])
+    str_x = np.lib.stride_tricks.as_strided(
+                gx,
+                shape=(bs, ctx.groups, cin, oy, ox, H, W),
+                strides=(gx.strides[0], gx.strides[1], gx.strides[2], gx.strides[3]*ys, gx.strides[4]*xs, gx.strides[3], gx.strides[4]),
+                writeable=False, 
+            )
+    return np.einsum('igjYXyx,gkjyx -> igkYX', str_x, w.reshape(ctx.groups,rcout, cin, H, W)).reshape(bs, cout, oy, ox)
```
Comment (geohot) [inline tinygrad/ops.py:174]: perhaps tw=w.reshape(ctx.groups,rcout, cin, H, W) by the gx line?

File: tinygrad/ops.py [hunk 2]
```diff
@@ -183,16 +182,29 @@ def backward(ctx, grad_output):
...
+    str_x = np.lib.stride_tricks.as_strided(
+                gx,
+                shape=(bs, ctx.groups, cin, oy, ox, H, W),
+                strides=(gx.strides[0], gx.strides[1], gx.strides[2], gx.strides[3]*ys, gx.strides[4]*xs, gx.strides[3], gx.strides[4]),
+                writeable=False, 
+            )
+
+    gdw = np.einsum('igkYX,igjYXyx -> gkjyx',ggg,str_x)
+   
+    #needs to be optimized
```
Comment (geohot) [inline tinygrad/ops.py:201]: Yea, I had trouble with this too when I wrote the fast convs. It's the stupid add accumulation I don't know how to do fast.

File: tinygrad/ops.py [hunk 3]
```diff
@@ -183,16 +182,29 @@ def backward(ctx, grad_output):
...
-      for Y in range(grad_output.shape[2]):
-        for X in range(grad_output.shape[3]):
+
+    gdx = dx.reshape(bs,ctx.groups,cin,x.shape[2],x.shape[3])
+    gdw = dw.reshape(ctx.groups,rcout, cin, H, W)
+
+    ggg = grad_output.reshape(grad_output.shape[0],ctx.groups,rcout,oy,ox)
+
+    gx = x.reshape(x.shape[0],ctx.groups,cin,x.shape[2],x.shape[3])
+    str_x = np.lib.stride_tricks.as_strided(
```
Comment (geohot) [inline tinygrad/ops.py:192]: Save this from the forward pass in the context instead of doing it twice?

File: tinygrad/ops.py [hunk 4]
```diff
@@ -183,16 +182,29 @@ def backward(ctx, grad_output):
     rcout = cout//ctx.groups
 
     dx, dw = np.zeros_like(x), np.zeros_like(w)
```
Comment (geohot) [inline tinygrad/ops.py:184]: This dw isn't used

**Comment by geohot** [comment]
OMG efficientnet is over 10x faster!

**Comment by geohot** [comment]
Actually I like this so much I'll merge and clean up

File: tinygrad/ops.py (hunk 1 above)
Comment (marcelbischoff) [reply tinygrad/ops.py:174]: I had it like this before

File: tinygrad/ops.py (hunk 2 above)
Comment (marcelbischoff) [reply tinygrad/ops.py:201]: probably can be done with strides, have to think, but you can not just einsum into a strided array (even if you could you shouldn't)

File: tinygrad/ops.py (hunk 4 above)
Comment (marcelbischoff) [reply tinygrad/ops.py:184]: my bad!

**Comment by marcelbischoff** [comment]
Here is my notebook code trying to understand how `as_strided()` works:
`import numpy as np

x = np.arange(12*12).reshape(12,12)
H, W, xs, ys =2, 3, 4, 5
oy, ox = (x.shape[0]-(H-ys))//ys, (x.shape[1]-(W-xs))//xs

str_x = np.lib.stride_tricks.as_strided(
    x,
    shape=(ox,oy,W,H),
    strides=(
        x.strides[0]*xs,
        x.strides[1]*ys,
        x.strides[0],
        x.strides[1],
    ),
    writeable=False
)`

**Comment by geohot** [comment]
Wait, this is a ton slower for batches!

forward pass:  1943.915 ms, 65.49x off baseline 29.681 ms
backward pass: 1650.604 ms, 43.55x off baseline 37.897 ms

**Comment by marcelbischoff** [comment]
Wondering if you have to put on einstein optimization. For the small size, it was slower with optimization. You could compute the path and store it. Haven't had time testing it...

**Comment by marcelbischoff** [comment]
I think it doesn't like the stride over the big array!

**Comment by geohot** [comment]
Yea, maybe reorder and put the batches last?

**Comment by marcelbischoff** [comment]
Maybe? So I think I get a feeling what is going on! einsum can be slower than matrix multiplication which is more specialized. For your example the X,Y,g loop only loops around 100 or 600 times, so then it seems faster to do matrix multiplication that often then doing the monstrous einsum. But somehow that should be possible to avoid...

**Comment by marcelbischoff** [comment]
that nested sum is faster 'ret = np.einsum('...yx -> ...',np.einsum('igjYXyx,gkjyx -> igkYXyx', tx, tw))' by 2

**Comment by marcelbischoff** [comment]
Okay, I think the reason is that einsum is so much slower than dot, tensordot which uses BLAS etc. How does `1.88x off baseline` in `test/test_net_speed.py` and `1.47 s` for `test/efficientnet.py` on a 1.2 GHz Dual-Core m3 sound?

**Comment by geohot** [comment]
That's not bad

**Comment by marcelbischoff** [comment]
oh didn't see the ravel, yeah it is not that bad withit, but doesn't this cause a memory issue? In tx a lot of indices point to the same value?

**Pull request merged**